### DISCUSSION TITLE
Series of Gleam Articles

### DISCUSSION THREAD

**Discussion by dusty-phillips** [gleam-lang/gleam#1618] [Show and tell]
I stumbled across Gleam a while ago and have found it very entertaining to write in. I noticed there isn't a ton of third-party content on it, so I started writing some. I don't have much Erlang or functional programming background, so I wouldn't take any of this as good practice, but hopefully some might find it interesting, and others might be able to provide some professional feedback:

* [Part 1: Hello World](https://dusty.phillips.codes/2022/04/17/a-tutorial-introduction-to-gleam-part-1/)
* [Part 2: Looping](https://dusty.phillips.codes/2022/05/04/a-tutorial-introduction-to-gleam-part-2/)

**Comment by michallepicki** [comment]
Looking forward to next articles! As I use functional programming day to day I find your insights on Gleam very interesting. For example, I rarely use the `iterator` module - I usually reach for the `map`, `fold`, `fold_until` and similar functions from the `list` module, or write recursion by hand. But I think it makes sense that the `iterator` module interface seems "closer" to for/while loops.

About the Erlang list representation and copying: data can be shared between lists when you're prepending to a list. If your list can be long, prefer prepend (`[new_elem1, new_elem2, ..old_list]` or [list.prepend](https://hexdocs.pm/gleam_stdlib/gleam/list.html#prepend)) instead of [list.append](https://hexdocs.pm/gleam_stdlib/gleam/list.html#append) (there's a note about performance in documentation there). A common pattern is to build up a list with prepending and [list.reverse](https://hexdocs.pm/gleam_stdlib/gleam/list.html#reverse) at the end.

**Reply by lpil** [reply]
I typically only use `iterator` if reading the entire input would use too much memory, and instead I wish to stream the input to the output and perform transformations as I go. For example processing lines in a large file.

**Reply by dusty-phillips** [reply]
I guess I defaulted to looking for iterator because that's what Python encourages. I'll try to do another article on the structure of lists later; I imagine it'll require some research.

**Comment by inoas** [comment]
Read through Part 1

I believe not having runtime exceptions is very high on the core dev/main devs list/s - so high in fact that 1/0 runs (because its types check out it has to :D) - so I am pretty certain full exhaustiveness checking will happen (other than say C target which onl hopefully will <3) - in short: I’d just write that it is not yet implemented but help is welcome :wink: - otherwise it looks like gleam does neglect that on by design imho (much like nulls in dart 1.0 :/)

I love the way piping / `|>` is explained to be read as `becomes the first argument of`.

`iodata` is now called `string_builder` AFAIK.

**Reply by lpil** [reply]
Yup, we will have full exhaustiveness checking #34

**Reply by dusty-phillips** [reply]
The iodata code was copied from the documentation. I put https://github.com/gleam-lang/website/pull/170 up to correct the docs if that is correct.

**Comment by lpil** [comment]
Wonderful stuff! Have you shared these on twitter, reddit, etc? I'd love to re-share them, I think people would enjoy them :)

Some notes on the first one:

> If you’re not using a Mac, you may have to go to a bit more effort, installing Erlang as a separate step and possibly building the Gleam binary from scratch. The current Gleam compiler is written in Rust, so you’ll have to install an entire separate programming language toolchain.

Rust isn't required to use Gleam, you can generally download a precompiled version from GitHub (https://github.com/gleam-lang/gleam/releases), and we're on a few of the various package managers.

> Note that in order to run this program you’ll need to use -- to tell gleam that your remaining arguments should go to your program, rather than to the gleam binary. So, for example, run gleam run -- Dusty instead of gleam run Dusty.

The `--` isn't required unless you are passing `--flags` to the Gleam program, and that is due to a bug in clap. Once they have fixed that `--` won't be required unless you're giving a flag that `gleam run` itself accepts.

> In Python, I’d write " ".join(names), but Gleam doesn’t have that (admittedly bizarre) construct.

Check out `string.join` https://hexdocs.pm/gleam_stdlib/gleam/string.html#join

**Reply by dusty-phillips** [reply]
Thank you! I'll make a few updates. I actually don't use twitter or reddit, so I've nothing to reshare, but feel free to link to it anywhere.!

**Comment by dusty-phillips** [comment]
I've created a couple more articles as I become more comfortable in the language:

* [Part 3: More on tail recursion](https://dusty.phillips.codes/2022/05/23/a-tutorial-introduction-to-gleam-part-3/)
* [Part 4: Introducing OTP](https://dusty.phillips.codes/2022/06/05/a-tutorial-introduction-to-gleam-part-4/)

**Reply by michallepicki** [reply]
I noticed links to previous articles lead to 404s, they might need small tweaks

**Reply by dusty-phillips** [reply]
Should be fixed now. The links Hugo generates when testing locally are different from those when it builds for production. Very strange.

**Reply by gofenix** [reply]
awesome

**Reply by TanklesXL** [reply]
Hey great articles! Just a note on this bit:
> Let’s now import gleam/crypto and calculate the hash of that password:   `crypto.hash(crypto.Sha256, password) |> io.debug`
This one line example demonstrates why we need to be especially careful of how we design APIs in Gleam. In my opinion, the hash function should accept the password bitstring as the first argument. Then I could have piped the password in as part of the original pipeline.

 What you want in terms of data-first piping in this case is actually perfectly doable by using positional piping with `_` as follows:
```gleam
  password
  |> crypto.hash(crypto.Sha256,_) 
  |> io.debug
```
Although I do agree for something like this the bit string should be the first argument, but it's nice that with `_` it doesn't make that much difference for an end-user of the library.

**Reply by dusty-phillips** [reply]
Oh wow, I can’t wait to try that, thanks!

**Comment by dusty-phillips** [comment]
Wrote an entire article inspired by @TanklesXL tip: https://dusty.phillips.codes/2022/07/24/a-tutorial-introduction-to-gleam-part-5/ Thank you!

**Reply by TanklesXL** [reply]
Oh wow, nicely done! 🎉 Glad to have been a source of inspiration
